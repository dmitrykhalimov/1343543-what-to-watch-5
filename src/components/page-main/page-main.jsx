import React from "react";
import {connect} from "react-redux";

import PropTypes from "prop-types";

import FilmsList from "../films-list/films-list";
import Filter from "../filter/filter";
import Footer from "../footer/footer";

import {changeGenre, incrementRendered, resetRendered} from "../../store/action";
import {validFilm, validUserData} from "../../utils/props";
import ShowMore from "../show-more/show-more";
import FilmsCatalog from "../films-catalog/films-catalog";
import PageContent from "../page-content/page-content";
import {Link} from "react-router-dom";
import {getRendered, getGenresList, getActiveGenre, getFilms, getFilteredFilms, getUserData} from "../../store/reducers/selectors";
import UserBlock from "../user-block/user-block";
import Logo from "../logo/logo";
import ButtonPlay from "../button-play/button-play";


const PageMain = (props) => {
  const {
    title,
    genre,
    year,
    films,
    activeGenre,
    filterChange,
    filteredFilms,
    genresList,
    rendered,
    userData,
    incrementRenderedFilms,
  } = props;

  return (
    <React.Fragment>
      <section className="movie-card">
        <div className="movie-card__bg">
          <img src="img/bg-the-grand-budapest-hotel.jpg" alt="The Grand Budapest Hotel" />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <header className="page-header movie-card__head">
          <Logo/>
          <UserBlock
            userData = {userData}
          />
        </header>

        <div className="movie-card__wrap">
          <div className="movie-card__info">
            <div className="movie-card__poster">
              <img src="img/the-grand-budapest-hotel-poster.jpg" alt="The Grand Budapest Hotel poster" width="218" height="327" />
            </div>

            <div className="movie-card__desc">
              <h2 className="movie-card__title">{title}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{genre}</span>
                <span className="movie-card__year">{year}</span>
              </p>

              <div className="movie-card__buttons">
                {/* TODO: Перерисовывается кнопка воспроизвести фильм, придумать чего делать с этим*/}
                <Link className="btn btn--play movie-card__button" type="button" to={`/films/1`}>
                  <ButtonPlay/>
                </Link>
                <button className="btn btn--list movie-card__button" type="button">
                  <svg viewBox="0 0 19 20" width="19" height="20">
                    <use xlinkHref="#add"></use>
                  </svg>
                  <span>My list</span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </section>

      <PageContent>
        <FilmsCatalog>
          <Filter
            genres = {genresList}
            activeGenre = {activeGenre}
            onFilterSelect = {filterChange}
          />

          <FilmsList
            films = {filteredFilms}
            maxQuantity = {rendered}
          />
          {filteredFilms.length > rendered ?
            <ShowMore
              rendered = {rendered}
              filmsQuantity = {films.length}
              onShowMore = {incrementRenderedFilms}
            /> : ``}
        </FilmsCatalog>
        <Footer
          isLight = {true}
        />
      </PageContent>
    </React.Fragment>
  );
};

PageMain.propTypes = {
  title: PropTypes.string.isRequired,
  genre: PropTypes.string.isRequired,
  year: PropTypes.string.isRequired,
  films: PropTypes.arrayOf(validFilm).isRequired,
  filteredFilms: PropTypes.arrayOf(validFilm).isRequired,
  activeGenre: PropTypes.string.isRequired,
  filterChange: PropTypes.func.isRequired,
  genresList: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  rendered: PropTypes.number.isRequired,
  incrementRenderedFilms: PropTypes.func.isRequired,
  userData: validUserData,
};

const mapStateToProps = (state) => ({
  activeGenre: getActiveGenre(state),
  films: getFilms(state),
  filteredFilms: getFilteredFilms(state),
  genresList: getGenresList(state),
  rendered: getRendered(state),
  userData: getUserData(state),
});

const mapDispatchToProps = (dispatch) => ({
  filterChange(genre) {
    dispatch(changeGenre(genre));
    dispatch(resetRendered());
  },

  incrementRenderedFilms(increment) {
    dispatch(incrementRendered(increment));
  }
});

// export {PageMain};
export default connect(mapStateToProps, mapDispatchToProps)(PageMain);
