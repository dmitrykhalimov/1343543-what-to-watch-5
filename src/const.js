export const ratingRanks = [
  {
    'Bad': 2.9
  },
  {
    'Normal': 4.9
  },
  {
    'Good': 7.9
  },
  {
    'Very good': 9.9
  },
  {
    'Awesome': 10
  }
];
