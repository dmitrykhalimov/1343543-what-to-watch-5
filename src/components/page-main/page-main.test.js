import React from "react";
import renderer from "react-test-renderer";
import {PageMain} from "./page-main";
import {TEST_MOCKS, TEST_MOCK_STORE} from "../../const";
import {MemoryRouter} from "react-router-dom";
import configureMockStore from "redux-mock-store";
import {Provider} from "react-redux";

const mockStore = configureMockStore();
const store = mockStore(TEST_MOCK_STORE);

describe(`PageMain`, () => {
  it(`Should PageMain render correctly`, () => {
    const tree = renderer
    .create(
        <Provider store={store}>
          <MemoryRouter>
            <PageMain
              films={TEST_MOCKS.films}
              filteredFilms={TEST_MOCKS.films}
              filmPromo={TEST_MOCKS.film}
              activeGenre={TEST_MOCKS.activeGenre}
              onFilterChange={TEST_MOCKS.noop}
              genresList={TEST_MOCKS.genres}
              rendered={8}
              onIncrementFilms={TEST_MOCKS.noop}
            />,
          </MemoryRouter>
        </Provider>
        , {
          createNodeMock: () => {
            return {};
          }
        }
    )
  .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
