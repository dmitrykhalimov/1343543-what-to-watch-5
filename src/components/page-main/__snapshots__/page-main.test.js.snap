// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageMain Should PageMain render correctly 1`] = `
Array [
  "",
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Interstate 60: Episodes of the Road"
        src="#978F7E"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          href="/mylist"
          onClick={[Function]}
        >
          <div
            className="user-block__avatar"
          >
            <img
              alt="User avatar"
              height="63"
              src="https://i.kym-cdn.com/photos/images/newsfeed/000/607/429/84b.jpg"
              width="63"
            />
          </div>
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="Interstate 60: Episodes of the Road"
            height="327"
            src="https://m.media-amazon.com/images/M/MV5BZDYwYzlhOTAtNDAwMC00ZTBhLWI4M2QtMTA1NmJhYTdiNTkxXkEyXkFqcGdeQXVyNTM0NTU5Mg@@._V1_UX182_CR0,0,182,268_AL_.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Interstate 60: Episodes of the Road
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Adeventure
            </span>
            <span
              className="movie-card__year"
            >
              2019
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              className="btn btn--play movie-card__button"
              href="/player/29"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#in-list"
                />
              </svg>
              <span>
                In list
              </span>
            </button>
            
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Comedy
          </a>
        </li>
        <li
          className="catalog__genres-item catalog__genres-item--active"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Adventure
          </a>
        </li>
        <li
          className="catalog__genres-item"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Crime
          </a>
        </li>
        <li
          className="catalog__genres-item"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Historical
          </a>
        </li>
        <li
          className="catalog__genres-item"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Thriller
          </a>
        </li>
        <li
          className="catalog__genres-item"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Drama
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            className="small-movie-card__link"
            href="/films/29"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <video
                autoPlay={false}
                height="175"
                muted={true}
                poster="https://thenewswheel.com/wp-content/uploads/2014/08/Interstate-60-review-BMW-convertible-James-Marsden-8-ball.jpg"
                src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Interstate 60: Episodes of the Road
              </span>
            </h3>
          </a>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            className="small-movie-card__link"
            href="/films/29"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <video
                autoPlay={false}
                height="175"
                muted={true}
                poster="https://thenewswheel.com/wp-content/uploads/2014/08/Interstate-60-review-BMW-convertible-James-Marsden-8-ball.jpg"
                src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Interstate 60: Episodes of the Road
              </span>
            </h3>
          </a>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            className="small-movie-card__link"
            href="/films/29"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <video
                autoPlay={false}
                height="175"
                muted={true}
                poster="https://thenewswheel.com/wp-content/uploads/2014/08/Interstate-60-review-BMW-convertible-James-Marsden-8-ball.jpg"
                src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Interstate 60: Episodes of the Road
              </span>
            </h3>
          </a>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            className="small-movie-card__link"
            href="/films/29"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <video
                autoPlay={false}
                height="175"
                muted={true}
                poster="https://thenewswheel.com/wp-content/uploads/2014/08/Interstate-60-review-BMW-convertible-James-Marsden-8-ball.jpg"
                src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Interstate 60: Episodes of the Road
              </span>
            </h3>
          </a>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            className="small-movie-card__link"
            href="/films/29"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <video
                autoPlay={false}
                height="175"
                muted={true}
                poster="https://thenewswheel.com/wp-content/uploads/2014/08/Interstate-60-review-BMW-convertible-James-Marsden-8-ball.jpg"
                src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
                width="280"
              />
            </div>
            <h3
              className="small-movie-card__title"
            >
              <span
                className="small-movie-card__link"
              >
                Interstate 60: Episodes of the Road
              </span>
            </h3>
          </a>
        </article>
      </div>
      
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
  ",",
]
`;
